### ISSUE TITLE
BlockingConnection.close() hangs on Exception treatment

### ISSUE THREAD

**Issue by galindro** [pika/pika#912]
The console hangs if you close a BlockingConnection on treating an exception triggered by exchange_declare() method from channel instance. To exit the program, you need to press Ctrl+C or close the connection from RabbitMQ Server.

## Env. Info:
* Pika version: 0.11.2
* Python version: 3.6.4
* RabbitMQ version: rabbitmq:3.7.2-management-alpine (docker image)

## How to reproduce
1. Start a rabbitmq docker container 
```bash
docker run -it --rm -p 15672:15672 -p 5672:5672 rabbitmq:3.7.2-management-alpine
```

2. Run the bellow script:
```python
import pika
import logging
import sys

logger = logging.getLogger()
h = logging.StreamHandler(sys.stdout)
FORMAT = f'%(asctime)s - %(levelname)s: "%(message)s"'
h.setFormatter(logging.Formatter(FORMAT))
logger.addHandler(h)
logger.setLevel("DEBUG")

rmq_credentials = pika.PlainCredentials('guest', 'guest')
rmq_params = pika.ConnectionParameters(host='localhost',
                                       port='5672',
                                       credentials=rmq_credentials,
                                       blocked_connection_timeout=5)
rmq_connection = pika.BlockingConnection(rmq_params)
rmq_channel = rmq_connection.channel()

try:
  rmq_channel.exchange_declare(exchange=1, exchange_type='topic', internal=True, durable=True)
except Exception as e:
  rmq_connection.close()
```

3. You should see the bellow logs and the console will hang. To exit, you will need to press Ctrl+C
```bash
2018-01-03 16:52:00,481 - DEBUG: "Using EPollPoller"
2018-01-03 16:52:00,481 - DEBUG: "Added: {'callback': <bound method Connection._on_connection_start of <SelectConnection CLOSED socket=None params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 1}"
2018-01-03 16:52:00,482 - DEBUG: "Added: {'callback': <bound method Connection._on_connection_close of <SelectConnection CLOSED socket=None params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 1}"
2018-01-03 16:52:00,482 - DEBUG: "Added: {'callback': <bound method Connection._on_connection_blocked of <SelectConnection CLOSED socket=None params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'one_shot': False, 'only': None, 'arguments': None}"
2018-01-03 16:52:00,482 - DEBUG: "Added: {'callback': <bound method Connection._on_connection_unblocked of <SelectConnection CLOSED socket=None params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'one_shot': False, 'only': None, 'arguments': None}"
2018-01-03 16:52:00,482 - DEBUG: "Added: {'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f97a5c42408>>, 'one_shot': False, 'only': None, 'arguments': None}"
2018-01-03 16:52:00,482 - DEBUG: "Added: {'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f97a6c2c948>>, 'one_shot': False, 'only': None, 'arguments': None}"
2018-01-03 16:52:00,482 - DEBUG: "Added: {'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f97a49d7048>>, 'one_shot': False, 'only': None, 'arguments': None}"
2018-01-03 16:52:00,482 - DEBUG: "add_timeout: added timeout -5511572320841547119; deadline=0 at 1515005520.4827719"
2018-01-03 16:52:00,484 - INFO: "Connecting to ::1:5672"
2018-01-03 16:52:00,486 - DEBUG: "Processing 0:Connection.Start"
2018-01-03 16:52:00,486 - DEBUG: "Processing use of oneshot callback"
2018-01-03 16:52:00,487 - DEBUG: "0 registered uses left"
2018-01-03 16:52:00,487 - DEBUG: "Removing callback #0: {'callback': <bound method Connection._on_connection_start of <SelectConnection PROTOCOL socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 0}"
2018-01-03 16:52:00,487 - DEBUG: "Calling <bound method Connection._on_connection_start of <SelectConnection PROTOCOL socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>> for "0:Connection.Start""
2018-01-03 16:52:00,487 - DEBUG: "Added: {'callback': <bound method Connection._on_connection_tune of <SelectConnection START socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 1}"
2018-01-03 16:52:00,488 - DEBUG: "Processing 0:Connection.Tune"
2018-01-03 16:52:00,488 - DEBUG: "Processing use of oneshot callback"
2018-01-03 16:52:00,488 - DEBUG: "0 registered uses left"
2018-01-03 16:52:00,488 - DEBUG: "Removing callback #0: {'callback': <bound method Connection._on_connection_tune of <SelectConnection START socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 0}"
2018-01-03 16:52:00,488 - DEBUG: "Calling <bound method Connection._on_connection_tune of <SelectConnection START socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>> for "0:Connection.Tune""
2018-01-03 16:52:00,488 - DEBUG: "Creating a HeartbeatChecker: 60"
2018-01-03 16:52:00,488 - DEBUG: "add_timeout: added timeout -4563005670168967375; deadline=60 at 1515005580.4888728"
2018-01-03 16:52:00,489 - DEBUG: "Added: {'callback': <bound method Connection._on_connection_open of <SelectConnection TUNE socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 1}"
2018-01-03 16:52:00,500 - DEBUG: "Processing 0:Connection.OpenOk"
2018-01-03 16:52:00,500 - DEBUG: "Processing use of oneshot callback"
2018-01-03 16:52:00,501 - DEBUG: "0 registered uses left"
2018-01-03 16:52:00,501 - DEBUG: "Removing callback #0: {'callback': <bound method Connection._on_connection_open of <SelectConnection TUNE socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 0}"
2018-01-03 16:52:00,501 - DEBUG: "Calling <bound method Connection._on_connection_open of <SelectConnection TUNE socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>> for "0:Connection.OpenOk""
2018-01-03 16:52:00,501 - DEBUG: "Processing 0:_on_connection_open"
2018-01-03 16:52:00,501 - DEBUG: "Calling <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f97a6c2c948>> for "0:_on_connection_open""
2018-01-03 16:52:00,501 - DEBUG: "Creating channel 1"
2018-01-03 16:52:00,501 - DEBUG: "Added: {'callback': <bound method Connection._on_channel_cleanup of <SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'one_shot': True, 'only': <Channel number=1 CLOSED conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>, 'arguments': None, 'calls': 1}"
2018-01-03 16:52:00,501 - DEBUG: "Added: {'callback': <bound method Channel._on_getempty of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'one_shot': False, 'only': None, 'arguments': None}"
2018-01-03 16:52:00,501 - DEBUG: "Added: {'callback': <bound method Channel._on_cancel of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'one_shot': False, 'only': None, 'arguments': None}"
2018-01-03 16:52:00,501 - DEBUG: "Added: {'callback': <bound method Channel._on_flow of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'one_shot': False, 'only': None, 'arguments': None}"
2018-01-03 16:52:00,502 - DEBUG: "Added: {'callback': <bound method Channel._on_close of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 1}"
2018-01-03 16:52:00,502 - DEBUG: "Entering blocking state on frame <Channel.Open(['out_of_band='])>; acceptable_replies=[<class 'pika.spec.Channel.OpenOk'>]"
2018-01-03 16:52:00,502 - DEBUG: "Adding on_synchronous_complete callback"
2018-01-03 16:52:00,502 - DEBUG: "Added: {'callback': <bound method Channel._on_synchronous_complete of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 1}"
2018-01-03 16:52:00,502 - DEBUG: "Adding passed-in callback"
2018-01-03 16:52:00,502 - DEBUG: "Added: {'callback': <bound method Channel._on_openok of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 1}"
2018-01-03 16:52:00,502 - DEBUG: "Added: {'callback': <bound method BlockingChannel._on_consumer_cancelled_by_broker of <BlockingChannel impl=<Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>>, 'one_shot': False, 'only': None, 'arguments': None}"
2018-01-03 16:52:00,502 - DEBUG: "Added: {'callback': <bound method _CallbackResult.signal_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f97a6a28108>>, 'one_shot': False, 'only': None, 'arguments': None}"
2018-01-03 16:52:00,502 - DEBUG: "Added: {'callback': <bound method BlockingChannel._on_channel_closed of <BlockingChannel impl=<Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 1}"
2018-01-03 16:52:00,503 - DEBUG: "Added: {'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f97a6a28188>>, 'one_shot': False, 'only': None, 'arguments': None}"
2018-01-03 16:52:00,503 - INFO: "Created channel=1"
2018-01-03 16:52:00,505 - DEBUG: "Processing 1:Channel.OpenOk"
2018-01-03 16:52:00,505 - DEBUG: "Processing use of oneshot callback"
2018-01-03 16:52:00,505 - DEBUG: "0 registered uses left"
2018-01-03 16:52:00,505 - DEBUG: "Removing callback #0: {'callback': <bound method Channel._on_synchronous_complete of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 0}"
2018-01-03 16:52:00,506 - DEBUG: "Processing use of oneshot callback"
2018-01-03 16:52:00,506 - DEBUG: "0 registered uses left"
2018-01-03 16:52:00,506 - DEBUG: "Removing callback #0: {'callback': <bound method Channel._on_openok of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 0}"
2018-01-03 16:52:00,506 - DEBUG: "Calling <bound method Channel._on_synchronous_complete of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>> for "1:Channel.OpenOk""
2018-01-03 16:52:00,506 - DEBUG: "0 blocked frames"
2018-01-03 16:52:00,506 - DEBUG: "Calling <bound method Channel._on_openok of <Channel number=1 OPENING conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>> for "1:Channel.OpenOk""
2018-01-03 16:52:00,506 - DEBUG: "Entering blocking state on frame <Exchange.Declare(['arguments={}', 'auto_delete=False', 'durable=True', 'exchange=1', 'internal=True', 'nowait=False', 'passive=False', 'ticket=0', 'type=topic'])>; acceptable_replies=[<class 'pika.spec.Exchange.DeclareOk'>]"
2018-01-03 16:52:00,506 - DEBUG: "Adding on_synchronous_complete callback"
2018-01-03 16:52:00,506 - DEBUG: "Added: {'callback': <bound method Channel._on_synchronous_complete of <Channel number=1 OPEN conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 1}"
2018-01-03 16:52:00,507 - DEBUG: "Adding passed-in callback"
2018-01-03 16:52:00,507 - DEBUG: "Added: {'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f97a6a28588>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 1}"
2018-01-03 16:52:00,507 - INFO: "Closing connection (200): Normal shutdown"
2018-01-03 16:52:00,507 - DEBUG: "Channel.close(200, Normal shutdown)"
2018-01-03 16:52:00,507 - DEBUG: "Added: {'callback': <bound method _CallbackResult.signal_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f97a6a28a08>>, 'one_shot': True, 'only': None, 'arguments': None, 'calls': 1}"
2018-01-03 16:52:00,507 - INFO: "Closing channel (200): 'Normal shutdown' on <Channel number=1 OPEN conn=<SelectConnection OPEN socket=('::1', 45152, 0, 0)->('::1', 5672, 0, 0) params=<ConnectionParameters host=localhost port=5672 virtual_host=/ ssl=False>>>"
2018-01-03 16:52:00,507 - DEBUG: "Already in blocking state, so enqueueing method <Channel.Close(['class_id=0', 'method_id=0', 'reply_code=200', 'reply_text=Normal shutdown'])>; acceptable_replies=[<class 'pika.spec.Channel.CloseOk'>]"
```

**Comment by vitaly-krugl** [comment]
@galindro, please post the entire traceback of the exception that you would see without calling `rmq_connection.close()`.

One of the things that surprises me is that I don't see a log entry resulting from the failure associated with the exception.

**Comment by vitaly-krugl** [comment]
@lukebakken, wasn't this fixed not long ago. I am struggling to find the relevant pull request.

**Comment by vitaly-krugl** [comment]
Closing since it was fixed by PR #991.
